File: leetcode-java/src/leetcode/patterns/SearchA2DMatrixTwo.java
package leetcode.patterns;

public class SearchA2DMatrixTwo {

    /**
     * 1 2  3  4
     * 5 6  7  8
     * 9 10 11 12
     *
     * search 6
     * */
    public static boolean searchMatrix(int[][] matrix, int target) {
        for (var row: matrix) {
            if (searchMatrix(row, target)) return true;
        }
        return false;
    }

    public static boolean searchMatrix(int[] rows, int target) {
        var lower = 0;
        var high = rows.length - 1;
        while (lower <= high) {
            var mid = lower + (high - lower) / 2;
            if (rows[mid] == target) {
                return true;
            }
            if (rows[mid] > target) {
                high = mid - 1;
            } else {
                lower = mid + 1;
            }
        }
        return false;
    }

    public static boolean searchMatrixTwo(int[][] matrix, int target) {
        var n = matrix.length;
        var m = matrix[0].length;

        var i = 0;
        var j = matrix[0].length - 1;
        while (i < n && j >= 0) {
            if(matrix[i][j]==target)
                return true;
            else if(matrix[i][j]>target) {
                j--;
            } else if(matrix[i][j]<target) {
                i++;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println(searchMatrixTwo(new int[][] {
                new int[]{1, 4},
                new int[]{2, 5}
        }, 6));
    }
}

File: leetcode-scala/src/leetcode/medium/SearchA2DMatrixTwo.scala
package leetcode.medium

import scala.annotation.tailrec

object SearchA2DMatrixTwo {
  def searchMatrix(matrix: Array[Array[Int]], target: Int): Boolean = {
    searchHelper(matrix, target, 0, matrix.apply(0).length - 1)
  }

  @tailrec
  def searchHelper(matrix: Array[Array[Int]], target: Int, i: Int, j: Int): Boolean = {
    if (j < 0 || i >= matrix.length)
      return false
    if (matrix.apply(i).apply(j) == target) {
      true
    } else if (matrix.apply(i).apply(j) > target) {
      searchHelper(matrix, target, i, j - 1)
    } else {
      searchHelper(matrix, target, i + 1, j)
    }
  }
}

object Main extends App {
  println(SearchA2DMatrixTwo.searchMatrix(Array(Array(1, 1)), 20))
}
